### PR TITLE
[LIVE-2497] - Bugfix: Handle reorg on Eth family sync

### PR DESCRIPTION
**Pull request by lambertkevin**
### 📝 Description

This PR adds a failsafe to Ethereum incremental synchronisation which is based on the latest "stable" block hash in the Account's operations history.
As of right now a block is considered safe based on an arbitrary value `SAFE_REORG_THRESHOLD` set at `80`, but it looks like even with this security we can end up with a block hash that doesn't exist on chain.
The failsafe will now make sure a block exists before trying an incremental sync from it, and if it doesn't it's going to do a full sync instead.

### ❓ Context

- **Impacted projects**: `live-common` 
- **Linked resource(s)**: https://ledgerhq.atlassian.net/browse/LIVE-2497 

### ✅ Checklist

- [x] **Test coverage** 
- [x] **Atomic delivery** 
- [x] **No breaking changes** 

### 📸 Demo

Can't really be demoed. But an integration test reproducing the issue has been added.

### 🚀 Expectations to reach

_Please make sure you follow these [**Important Steps**](https://github.com/LedgerHQ/ledger-live/blob/develop/CONTRIBUTING.md#important-steps)._

_Pull Requests must pass the CI and be internally validated in order to be merged._

### CODE REVIEW

**Review by gre — CHANGES_REQUESTED** [review]

File: libs/ledger-live-common/src/families/ethereum/synchronisation.ts [hunk 1]
```diff
@@ -42,6 +42,10 @@ export const getAccountShape: GetAccountShape = async (
   // fetch transactions, incrementally if possible
   const mostRecentStableOperation = initialStableOperations[0];
   // when new tokens are added / blacklist changes, we need to sync again because we need to go through all operations again
+  // Check if the block hash exists on chain to prevent reorg issue
+  const blockHashExistsOnChain = await api
+    .getBlockByHash(mostRecentStableOperation?.blockHash)
+    .then(Boolean);
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply libs/ledger-live-common/src/families/ethereum/synchronisation.ts:48]: Done 👌

File: libs/ledger-live-common/src/bridge/jsHelpers.ts [hunk 2]
```diff
@@ -205,10 +206,16 @@ export const makeSync =
...
             syncConfig
           );
+
           o.next((acc) => {
+            const { family } = getCryptoCurrencyById(initial.currency.id);
             const a = needClear ? clearAccount(acc) : acc;
             // FIXME reconsider doing mergeOps here. work is redundant for impl like eth
-            const operations = mergeOps(a.operations, shape.operations || []);
+            const operations =
+              family === "ethereum"
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply libs/ledger-live-common/src/bridge/jsHelpers.ts]: Ok I introduced a new parameter to `makeSync` with `shouldMergeOps` with a default behaviour at `true` to ensure compatibility with the other coins.

I'm not really fond of adding new layers of compatibility without it being very visible and logical to mostCases, but I guess it's still a way better option than my "it ethereum" 👍

**Review by haammar-ledger — APPROVED** [review]
LGTM

File: libs/ledger-live-common/src/bridge/jsHelpers.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (haammar-ledger) [reply libs/ledger-live-common/src/bridge/jsHelpers.ts]: @gre please confirm if the proposed solution is acceptable to you.

**Review by gre — CHANGES_REQUESTED** [review]

File: libs/ledger-live-common/src/families/ethereum/test-dataset.ts [hunk 3]
```diff
@@ -30,6 +30,25 @@ export const ethereum1: AccountRaw = {
...
+  freshAddresses: [],
+  pendingOperations: [],
+  operations: [],
+  currencyId: "ethereum",
+  unitMagnitude: 18,
+  balance: "",
+  blockHeight: 0,
+  lastSyncDate: "",
+  xpub: "",
+};
```
Comment (gre) [inline libs/ledger-live-common/src/families/ethereum/test-dataset.ts:50]: can you move this directly in synchronisation.integration.test.ts ? this test-dataset.ts file is soon gone.

**Comment by gre** [comment]
https://github.com/LedgerHQ/ledger-live/actions/runs/2568591731

**Comment by lambertkevin** [comment]
Just for further investigation, here is a problematic transactions generating subOperations with the same ids that are now mutated by `mergeOps` when not adding the index in the subOperation id:

```json
{
  "hash": "0xe715b4530e40a59f9c943089ff80fc57415687daded99d289574c3dd6e125c4b",
  "transaction_type": 0,
  "nonce": "0x10",
  "value": 0,
  "gas": 334044,
  "gas_price": 5110000001,
  "max_fee_per_gas": null,
  "max_priority_fee_per_gas": null,
  "from": "0x58e327d64fc77fa36f698989623c3e4f058c8a89",
  "to": "0x4f833a24e1f95d70f028921e27040ca56e09ab0b",
  "transfer_events": {
      "list": [
          {
              "contract": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
              "from": "0x0e3f0bb9516f01f2c34c25e0957518b8ac9414c5",
              "to": "0x6924a03bb710eaf199ab6ac9f2bb148215ae9b5d",
              "count": 37531836988865469
          },
          {
              "contract": "0xd26114cd6ee289accf82350c8d8487fedb8a0c07",
              "from": "0x6924a03bb710eaf199ab6ac9f2bb148215ae9b5d",
              "to": "0x0e3f0bb9516f01f2c34c25e0957518b8ac9414c5",
              "count": 5000000000000000000
          },
          {
              "contract": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
              "from": "0x0e3f0bb9516f01f2c34c25e0957518b8ac9414c5",
              "to": "0x58e327d64fc77fa36f698989623c3e4f058c8a89",
              "count": 1752163011134531
          }
      ],
      "truncated": false,
      "next_index": null
  },
  "erc721_transfer_events": [],
  "erc1155_transfer_events": [],
  "approval_events": [],
  "actions": [
      {
          "from": "0x58e327d64fc77fa36f698989623c3e4f058c8a89",
          "to": "0x4f833a24e1f95d70f028921e27040ca56e09ab0b",
          "input": null,
          "value": 0,
          "gas": 282628,
          "gas_used": 171280,
          "error": null
      },
      {
          "from": "0x4f833a24e1f95d70f028921e27040ca56e09ab0b",
          "to": "0x2240dab907db71e64d3e0dba4800c83b5c502d4e",
          "input": null,
          "value": 0,
          "gas": 169859,
          "gas_used": 15079,
          "error": null
      },
      {
          "from": "0x2240dab907db71e64d3e0dba4800c83b5c502d4e",
          "to": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
          "input": null,
          "value": 0,
          "gas": 165834,
          "gas_used": 13625,
          "error": null
      },
      {
          "from": "0x4f833a24e1f95d70f028921e27040ca56e09ab0b",
          "to": "0x2240dab907db71e64d3e0dba4800c83b5c502d4e",
          "input": null,
          "value": 0,
          "gas": 153405,
          "gas_used": 20870,
          "error": null
      },
      {
          "from": "0x2240dab907db71e64d3e0dba4800c83b5c502d4e",
          "to": "0xd26114cd6ee289accf82350c8d8487fedb8a0c07",
          "input": null,
          "value": 0,
          "gas": 149637,
          "gas_used": 19416,
          "error": null
      },
      {
          "from": "0x4f833a24e1f95d70f028921e27040ca56e09ab0b",
          "to": "0x2240dab907db71e64d3e0dba4800c83b5c502d4e",
          "input": null,
          "value": 0,
          "gas": 131263,
          "gas_used": 15079,
          "error": null
      },
      {
          "from": "0x2240dab907db71e64d3e0dba4800c83b5c502d4e",
          "to": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
          "input": null,
          "value": 0,
          "gas": 127841,
          "gas_used": 13625,
          "error": null
      }
  ],
  "confirmations": 7017716,
  "input": null,
  "gas_used": 207696,
  "cumulative_gas_used": 7559212,
  "status": 1,
  "received_at": "2019-06-25T15:51:05Z",
  "block": {
      "hash": "0x441327cd5cb1def0af67178078788ae08715f0a46260ea5e92acf7b73c728fcf",
      "height": 8028070,
      "time": "2019-06-25T15:51:05Z"
  }
}
```

**Review by gre — APPROVED** [review]

**Comment by lambertkevin** [comment]
Except for a timeout issue on the integration test, we're good. `feature ready` 🙆‍♂️

**Pull request merged**